README.md
# open_ttt_lib

[![Crates.io](https://img.shields.io/crates/v/open_ttt_lib.svg)](https://crates.io/crates/open_ttt_lib)
[![Documentation](https://docs.rs/open_ttt_lib/badge.svg)](https://docs.rs/open_ttt_lib)
[![License](https://img.shields.io/crates/l/open_ttt_lib.svg)](https://github.com/j-richey/open_ttt_lib/blob/main/LICENSE.txt)
[![Build Status](https://travis-ci.com/j-richey/open_ttt_lib.svg?branch=main)](https://travis-ci.com/j-richey/open_ttt_lib)
[![Coverage Status](https://coveralls.io/repos/github/j-richey/open_ttt_lib/badge.svg?branch=main)](https://coveralls.io/github/j-richey/open_ttt_lib?branch=main)
[![unsafe forbidden](https://img.shields.io/badge/unsafe-forbidden-success.svg)](https://github.com/rust-secure-code/safety-dance/)

Open source Rust library containing common Tic Tac Toe functionality.

## Overview
Tic Tac Toe is a game of strategy where two players, X and O, take turns placing their mark in a 3 x 3 gird.
The first player to get three marks in a row, column, or diagonal wins the game. 
The game can also end in a draw, known as a *cat's game*.

This library contains logic to enforce the rules of Tic Tac Toe, manage the game's state,
and provides artificial intelligence algorithms for single player games.


## Usage
Add this to your `Cargo.toml`:

```toml
[dependencies]
open_ttt_lib = "0.2.2"
```

See the library's [documentation](https://docs.rs/open_ttt_lib/) for complete
details on the library's API.


## Examples
Below is a short example of using this library.

```rust
use open_ttt_lib::{ai, board, game};

fn main() -> Result<(), Box<game::Error>> {
    // Create a game struct to manage the game.
    let mut game = game::Game::new();

    // Pick a position to place a mark. Positions are zero based.
    // An error result is returned if the position is outside the bounds
    // of the board, the position is already owned, or the game is over.
    let position = board::Position { row: 0, column: 2 };
    game.do_move(position)?;

    // do_move() updates the game state. The state indicates the player
    // who gets to place to next mark or, if the game is over, the
    // outcome of the game.
    match game.state() {
        game::State::PlayerXMove => println!("X's turn."),
        game::State::PlayerOMove => println!("O's turn."),
        game::State::PlayerXWin(_) => println!("Game Over: X wins!"),
        game::State::PlayerOWin(_) => println!("Game Over: O wins!"),
        game::State::CatsGame => println!("Game Over: cat's game."),
    };

    // Have an AI opponent pick a move.
    let opponent = ai::Opponent::new(ai::Difficulty::Medium);
    if let Some(ai_position) = opponent.get_move(&game) {
        game.do_move(ai_position)?;
    };

    Ok(())
}
```

The `examples` directory contains additional examples.

To run the examples, clone this repository then use `cargo run --example`. E.g:
```text
git clone https://github.com/j-richey/open_ttt_lib.git
cd open_ttt_lib
cargo run --example single_player
```


## Benchmarks
This library includes benchmarks that you can use to evaluate if the library
fits in with your performance goals. Use `cargo bench` to run the benchmark
suite:

```text
git clone https://github.com/j-richey/open_ttt_lib.git
cd open_ttt_lib
cargo bench
```


## Reporting Issues and Feature Requests
Please report issues and feel free to request new features using this project's
[GitHub issue tracker](https://github.com/j-richey/open_ttt_lib/issues).


## Changes
All notable changes are documented in the [CHANGELOG.md](https://github.com/j-richey/open_ttt_lib/blob/main/CHANGELOG.md).


## Contributing
Contributions are welcome! See [CONTRIBUTING.md](https://github.com/j-richey/open_ttt_lib/blob/main/CONTRIBUTING.md)
for details on how to contribute to this project.


## License
The library is licensed under the [MIT license](https://github.com/j-richey/open_ttt_lib/blob/main/LICENSE.txt).


CHANGELOG.md
# Change Log
All notable changes to this project are documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
as described for Rust libraries in
[RFC 1105](https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md).


## [0.2.2] - 2021-05-22

### Fixed
* Several documentation typos.

### Changed
* Improved `game::Game::do_move()` speed by up to 70%.
* Improved the worst case AI move time by up to 60%.


## [0.2.1] - 2021-02-15

### Added
* Documented the library does not use unsafe code. This is enforced with the
  `#![forbid(unsafe_code)]` attribute. 

### Fixed
* Several documentation typos.


## [0.2.0] - 2020-07-18

### Added
* `ai::Difficulty` enumeration containing predefined AI difficulty levels
  including **Easy**, **Medium**, and **Hard** . The **Custom** variant allows
  full control over the AI opponent's difficulty.
* `ai_difficulties` example shows how the different AI difficulties compare and
  demonstrates how to create custom difficulties.

### Changed
* The `ai::Opponent::new` constructor takes a `ai::Difficulty` variant instead
  of a `mistake_probability` number.
* Improved the worst case AI move time by 150x.


## [0.1.2] - 2020-04-18

### Added
* `single_player` example that shows how to use the library to create a complete
  console based Tic Tac Toe game.
* Include a brief overview on the rules of Tic Tac Toe the library documentation.


## [0.1.1] - 2019-12-22

### Fixed
* The crate's README file not being displayed on crates.io


## [0.1.0] - 2019-12-21
First release!

## Added
* `game::Game` structure for handling management of Tic Tac Toe games.
* `ai::Opponent` structure that provides a computer controlled AI opponent.
* `board` module containing helper types for working with Tic Tac Toe boards.


tests/ai_tests.rs
use open_ttt_lib::{ai, game};

// Ensures the unbeatable opponent never loses.
//
// The unbeatable opponent fully evaluates every possible move and countermove
// to pick the best position. However, if there is an issue, it might lose to a
// random opponent. This test plays several games to ensure the unbeatable
// opponent does not lose.
//
// This test could take a while to exercise so it is disabled by default.
#[test]
#[ignore]
fn unbeatable_opponent_should_never_lose() {
    // The number of games to play. A larger number makes the test take longer
    // to run, but due to the random nature of the test, more likely to find
    // any possible issues.
    const NUM_GAMES: usize = 100;

    let mut game = game::Game::new();

    let random_ai = ai::Opponent::new(ai::Difficulty::None);
    let unbeatable_ai = ai::Opponent::new(ai::Difficulty::Unbeatable);

    for _ in 0..NUM_GAMES {
        let mut move_log = Vec::new();
        loop {
            match game.state() {
                game::State::PlayerXMove => {
                    let position = random_ai.get_move(&game).unwrap();
                    move_log.push(format!("  Random AI as X: {:?}", position));
                    game.do_move(position).unwrap();
                }
                game::State::PlayerOMove => {
                    let position = unbeatable_ai.get_move(&game).unwrap();
                    move_log.push(format!("  Unbeatable AI as O: {:?}", position));
                    game.do_move(position).unwrap();
                }
                game::State::PlayerXWin(_) => {
                    panic!(
                        "\nThe random AI has won over the unbeatable AI. \
                        \n\nList of moves: \n{}\n \
                        \nThe final game board: \n{}\n",
                        move_log.join("\n"),
                        game.board()
                    );
                }
                game::State::PlayerOWin(_) => {
                    break;
                }
                game::State::CatsGame => {
                    break;
                }
            };
        }
        game.start_next_game();
    }
}

// Ensures the easy, medium, and hard are progressively harder by battling a
// a random opponent. We expect the harder difficulties to win more often than
// the easier ones.
#[test]
#[ignore]
fn easy_medium_hard_difficulties_should_increasingly_win_vs_none_difficulty() {
    let easy_scores = battle(ai::Difficulty::Easy, ai::Difficulty::None);
    let medium_scores = battle(ai::Difficulty::Medium, ai::Difficulty::None);
    let hard_scores = battle(ai::Difficulty::Hard, ai::Difficulty::None);

    assert!(
        easy_scores.wins < medium_scores.wins,
        "The Easy difficulty with {} wins has unexpectedly won more than the \
         Medium difficulty with {} wins.",
        easy_scores.wins,
        medium_scores.wins
    );
    assert!(
        medium_scores.wins < hard_scores.wins,
        "The Medium difficulty with {} wins has unexpectedly won more than the \
         Hard difficulty with {} wins.",
        medium_scores.wins,
        hard_scores.wins
    );
}

// Ensures the easy, medium, and hard are progressively harder by battling a
// an unbeatable opponent. Because the unbeatable opponent is, well unbeatable,
// we expect the harder difficulties to get more cat's games than the easier
// opponents.
#[test]
#[ignore]
fn easy_medium_hard_difficulties_should_increasingly_tie_vs_unbeatable_difficulty() {
    let easy_scores = battle(ai::Difficulty::Easy, ai::Difficulty::Unbeatable);
    let medium_scores = battle(ai::Difficulty::Medium, ai::Difficulty::Unbeatable);
    let hard_scores = battle(ai::Difficulty::Hard, ai::Difficulty::Unbeatable);

    assert!(
        easy_scores.cats_games < medium_scores.cats_games,
        "The Easy difficulty with {} cat's games has unexpectedly tied more \
         than the Medium difficulty with {} cat's games.",
        easy_scores.cats_games,
        medium_scores.cats_games
    );
    assert!(
        medium_scores.cats_games < hard_scores.cats_games,
        "The Medium difficulty with {} cat's games has unexpectedly tied more \
        than the Hard difficulty with {} cat's games.",
        medium_scores.cats_games,
        hard_scores.cats_games
    );

    // Also have a sanity check that during the battle, the unbeatable was in
    // fact unbeaten. If these asserts fail then consider seeing if the
    // unbeatable_opponent_should_never_lose test can recreate the failure.
    assert_eq!(
        easy_scores.wins, 0,
        "The Easy opponent with {} wins has unexpectedly won over the \
         unbeatable opponent.",
        easy_scores.wins
    );
    assert_eq!(
        medium_scores.wins, 0,
        "The Medium opponent with {} wins has unexpectedly won over the \
         Unbeatable opponent.",
        medium_scores.wins
    );
    assert_eq!(
        hard_scores.wins, 0,
        "The Hard opponent with {} wins has unexpectedly won over the \
         Unbeatable opponent.",
        hard_scores.wins
    );
}

fn battle(difficulty: ai::Difficulty, reference_difficulty: ai::Difficulty) -> BattleScores {
    // The number of games to play in a battle. A larger number makes the test
    // take longer to run, but due to the random nature of the test, more
    // likely to find any possible issues.
    const NUM_GAMES: i32 = 100;

    // The game logic ensures each opponent takes turns taking the first move,
    // thus start_next_game() is used instead of creating a new game once the
    // game is over.
    let mut game = game::Game::new();
    let player_x = ai::Opponent::new(difficulty);
    let player_o = ai::Opponent::new(reference_difficulty);
    let mut scores = BattleScores::new();

    while scores.total_games() < NUM_GAMES {
        match game.state() {
            game::State::PlayerXMove => {
                let position = player_x.get_move(&game).unwrap();
                game.do_move(position).unwrap();
            }
            game::State::PlayerOMove => {
                let position = player_o.get_move(&game).unwrap();
                game.do_move(position).unwrap();
            }
            game::State::PlayerXWin(_) => {
                scores.wins += 1;
                game.start_next_game();
            }
            game::State::PlayerOWin(_) => {
                scores.losses += 1;
                game.start_next_game();
            }
            game::State::CatsGame => {
                scores.cats_games += 1;
                game.start_next_game();
            }
        };
    }

    scores
}

struct BattleScores {
    wins: i32,
    losses: i32,
    cats_games: i32,
}

impl BattleScores {
    fn new() -> Self {
        BattleScores {
            wins: 0,
            losses: 0,
            cats_games: 0,
        }
    }

    fn total_games(&self) -> i32 {
        self.wins + self.losses + self.cats_games
    }
}


tests/regression.rs
use open_ttt_lib::{ai, game};

// Ensures the unbeatable opponent never loses.
//
// The unbeatable opponent fully evaluates every possible move and countermove
// to pick the best position. However, if there is an issue, it might lose to a
// random opponent. This test plays several games to ensure the unbeatable
// opponent does not lose.
//
// This test could take a while to exercise so it is disabled by default.
#[test]
#[ignore]
fn unbeatable_position_should_never_lose() {
    // The number of games to play. A larger number makes the test take longer
    // to run, but due to the random nature of the test, more likely to find
    // any possible issues.
    const NUM_GAMES: usize = 100;

    let mut game = game::Game::new();

    let random_ai = ai::Opponent::new(ai::Difficulty::None);
    let unbeatable_ai = ai::Opponent::new(ai::Difficulty::Unbeatable);

    for _ in 0..NUM_GAMES {
        let mut move_log = Vec::new();
        loop {
            match game.state() {
                game::State::PlayerXMove => {
                    let position = random_ai.get_move(&game).unwrap();
                    move_log.push(format!("  Random AI as X: {:?}", position));
                    game.do_move(position).unwrap();
                }
                game::State::PlayerOMove => {
                    let position = unbeatable_ai.get_move(&game).unwrap();
                    move_log.push(format!("  Unbeatable AI as O: {:?}", position));
                    game.do_move(position).unwrap();
                }
                game::State::PlayerXWin(_) => {
                    panic!(
                        "\nThe random AI has won over the unbeatable AI. \
                        \n\nList of moves: \n{}\n \
                        \nThe final game board: \n{}\n",
                        move_log.join("\n"),
                        game.board()
                    );
                }
                game::State::PlayerOWin(_) => {
                    break;
                }
                game::State::CatsGame => {
                    break;
                }
            };
        }
        game.start_next_game();
    }
}


tests/version-numbers.rs
#[test]
fn readme_deps_version_number_should_match_cargo_toml() {
    version_sync::assert_markdown_deps_updated!("README.md");
}

#[test]
fn html_root_url_version_number_should_match_cargo_toml() {
    version_sync::assert_html_root_url_updated!("src/lib.rs");
}


src/lib.rs
//! # Overview
//! Tic Tac Toe is a game of strategy where two players, X and O, take turns
//! placing their mark in a 3 x 3 gird. The first player to get three marks in a
//! row, column, or diagonal wins the game. The game can also end in a draw,
//! known as a *cat's game*.
//!
//! This library contains logic to enforce the rules of Tic Tac Toe, manage the
//! game's state, and provides artificial intelligence algorithms for single
//! player games.
//!
//! # Usage
//! The [`Game`](game/struct.Game.html) structure is the central type provided
//! by this crate. It enforces the rules and manages the
//! [`State`](game/enum.State.html) of the game.
//! The [`Opponent`](ai/struct.Opponent.html) structure provides support for
//! single player games.
//!
//! # Example
//! ```
//! use open_ttt_lib::{ai, board, game};
//!
//! fn main() -> Result<(), Box<game::Error>> {
//!     // Create a game struct to manage the game.
//!     let mut game = game::Game::new();
//!
//!     // Pick a position to place a mark. Positions are zero based.
//!     // An error result is returned if the position is outside the bounds
//!     // of the board, the position is already owned, or the game is over.
//!     let position = board::Position { row: 0, column: 2 };
//!     game.do_move(position)?;
//!
//!     // do_move() updates the game state. The state indicates the player
//!     // who gets to place to next mark or, if the game is over, the
//!     // outcome of the game.
//!     match game.state() {
//!         game::State::PlayerXMove => println!("X's turn."),
//!         game::State::PlayerOMove => println!("O's turn."),
//!         game::State::PlayerXWin(_) => println!("Game Over: X wins!"),
//!         game::State::PlayerOWin(_) => println!("Game Over: O wins!"),
//!         game::State::CatsGame => println!("Game Over: cat's game."),
//!     };
//!
//!     // Have an AI opponent pick a move.
//!     let opponent = ai::Opponent::new(ai::Difficulty::Medium);
//!     if let Some(ai_position) = opponent.get_move(&game) {
//!         game.do_move(ai_position)?;
//!     };
//!
//!     Ok(())
//! }
//! ```

#![doc(html_root_url = "https://docs.rs/open_ttt_lib/0.2.2")]
#![forbid(unsafe_code)]

pub mod ai;
pub mod board;
pub mod game;

// Ensure the examples in the README file also work as expected.
extern crate doc_comment;
doc_comment::doctest!("../README.md", examples_in_readme);


benches/benchmarks.rs
#[macro_use]
extern crate criterion;

use criterion::black_box;
use criterion::Criterion;

use open_ttt_lib::ai;
use open_ttt_lib::game;

// Sequence of positions that results in a cats game.
// The resulting game board is as follows:
//  +---+---+---+
//  | X | O | X |
//  +---+---+---+
//  | X | O | O |
//  +---+---+---+
//  | O | X | X |
//  +---+---+---+
// Note: for a particular game the X and O spots might be reversed.
const CATS_GAME_POSITION_SEQUENCE: [game::Position; 9] = [
    game::Position { row: 0, column: 0 },
    game::Position { row: 0, column: 1 },
    game::Position { row: 0, column: 2 },
    game::Position { row: 1, column: 1 },
    game::Position { row: 1, column: 0 },
    game::Position { row: 1, column: 2 },
    game::Position { row: 2, column: 1 },
    game::Position { row: 2, column: 0 },
    game::Position { row: 2, column: 2 },
];

// Plays a complete game that results in a cat's game.
// The exercises the speed of the game's state machine and victory condition logic.
fn complete_game_benchmark(c: &mut Criterion) {
    let mut game = game::Game::new();

    c.bench_function("Complete game resulting in cats game.", |b| {
        b.iter(|| {
            for position in CATS_GAME_POSITION_SEQUENCE.iter() {
                game.do_move(black_box(*position)).unwrap();
            }
            game.start_next_game();
        })
    });
}

// Creates an unbeatable AI opponent then benchmarks for various numbers of free
// spaces remaining.
fn perfect_ai_moves_benchmarks(c: &mut Criterion) {
    let mut game = game::Game::new();

    let ai_opponent = ai::Opponent::new(ai::Difficulty::Unbeatable);

    // Loop through each position first benchmarking how long the AI takes to
    // select a position, doing the actual move with the predetermined position
    // so next time through the loop there are less free moves remaining.
    for idx in 0..CATS_GAME_POSITION_SEQUENCE.len() - 1 {
        let moves_remaining = game.free_positions().count();

        c.bench_function(
            &format!("Perfect AI with {} moves remaining", moves_remaining),
            |b| b.iter(|| ai_opponent.get_move(&game)),
        );

        game.do_move(CATS_GAME_POSITION_SEQUENCE[idx]).unwrap();
    }
}

criterion_group!(game_bench, complete_game_benchmark);

criterion_group!(
    name = perfect_ai_bench;
    config = Criterion::default().sample_size(50);
    targets = perfect_ai_moves_benchmarks);

criterion_main!(game_bench, perfect_ai_bench);


examples/ai_difficulties.rs
//! Example showing the different AI difficulties.

use rand::Rng;
use std::fmt;
use std::io;
use std::io::prelude::*;
use std::time;

use open_ttt_lib::{ai, game};

const INSTRUCTIONS: &str = r#"
AI Difficulty Examples
======================

This example shows how the different AI difficulties compare. AI opponents using
various difficulties play a series of games. The generated table shows the
percentage of wins, losses, and cat's games for each difficulty compared to the
None difficulty which places marks randomly and the Unbeatable difficulty which
never makes a mistake.

This example also demonstrates how to create custom difficulties. Try modifying
the `should_evaluate_node()` function and see how it compares to the builtin
difficulties.

Note: this example runs significantly faster with the --release flag: e.g:
$ cargo run --release --example ai_difficulties
"#;

// The number of games to play for each battle. More games gives a more accurate
// representation of how the difficulties compare, but takes longer to run.
const NUM_GAMES: i32 = 100;

// Custom difficulty's should evaluate node function. Modify this function to
// experiment with custom difficulties.
fn should_evaluate_node(depth: i32) -> bool {
    if depth == 0 {
        true
    } else {
        let evaluate_node_probability = 0.8;
        rand::thread_rng().gen_bool(evaluate_node_probability)
    }
}

fn main() {
    println!("{}", INSTRUCTIONS);
    print_table_header();

    evaluate_difficulty(ai::Difficulty::None);
    evaluate_difficulty(ai::Difficulty::Easy);
    evaluate_difficulty(ai::Difficulty::Medium);
    evaluate_difficulty(ai::Difficulty::Hard);
    evaluate_difficulty(ai::Difficulty::Custom(should_evaluate_node));
    evaluate_difficulty(ai::Difficulty::Unbeatable);
}

// Compares the provided difficulty to the reference difficulties. The results
// are printed to the screen.
fn evaluate_difficulty(difficulty: ai::Difficulty) {
    let difficulty_name = get_difficulty_name(&difficulty);

    let none_scores = battle(difficulty, ai::Difficulty::None);
    let unbeatable_scores = battle(difficulty, ai::Difficulty::Unbeatable);

    print_table_row(
        difficulty_name,
        &none_scores.to_string(),
        &unbeatable_scores.to_string(),
    );
}

// Has AI opponents of the provided difficulties play a series of games counting
// the wins for each player. Depending on the number of games being played, this
// function might take a while, so the progress of the battle is occasionally
// printed.
fn battle(
    player_x_difficulty: ai::Difficulty,
    player_o_difficulty: ai::Difficulty,
) -> BattleScores {
    // The game logic ensures each opponent takes turns taking the first move,
    // thus start_next_game() is used instead of creating a new game once the
    // game is over.
    let mut game = game::Game::new();

    let player_x_name = get_difficulty_name(&player_x_difficulty);
    let player_x = ai::Opponent::new(player_x_difficulty);
    let player_o_name = get_difficulty_name(&player_o_difficulty);
    let player_o = ai::Opponent::new(player_o_difficulty);

    let mut scores = BattleScores::new();

    let mut last_print_progress_time = time::Instant::now();

    while scores.total_games() < NUM_GAMES {
        // Play one turn of the either getting asking one of the AI players to
        // pick a position or if the game is over updating the scores and starting
        // the next game.
        match game.state() {
            game::State::PlayerXMove => {
                let position = player_x.get_move(&game).unwrap();
                game.do_move(position).unwrap();
            }
            game::State::PlayerOMove => {
                let position = player_o.get_move(&game).unwrap();
                game.do_move(position).unwrap();
            }
            game::State::PlayerXWin(_) => {
                scores.player_x_wins += 1;
                game.start_next_game();
            }
            game::State::PlayerOWin(_) => {
                scores.player_o_wins += 1;
                game.start_next_game();
            }
            game::State::CatsGame => {
                scores.cats_games += 1;
                game.start_next_game();
            }
        };

        print_battle_progress(
            scores.total_games(),
            player_x_name,
            player_o_name,
            &mut last_print_progress_time,
        );
    }

    scores
}

// Prints the table's header.
fn print_table_header() {
    println!("{:10}  {:^18}  {:^18}", "Difficulty", "None", "Unbeatable");
    println!("{:=<10}  {:=<18}  {:=<18}", "", "", "");
}

// Prints a row of the table.
fn print_table_row(col_1: &str, col_2: &str, col_3: &str) {
    println!("{:10}  {:18}  {:18}", col_1, col_2, col_3);
}

// Occasionally prints the progress of a battle.
fn print_battle_progress(
    games_played: i32,
    player_x_name: &str,
    player_o_name: &str,
    last_update_time: &mut time::Instant,
) {
    // The time between updates is set so users can see the program is making
    // progress but so it does not go so fast that the display is just a blur.
    const UPDATE_INTERVAL: time::Duration = time::Duration::from_millis(100);

    if last_update_time.elapsed() >= UPDATE_INTERVAL {
        // Create a description of the progress using the player names abd
        // number of games played.
        let battle_progress = games_played as f64 / NUM_GAMES as f64;
        let progress_text = format!(
            "{} vs. {}  game {} of {}, ({:.0}%)",
            player_x_name,
            player_o_name,
            games_played,
            NUM_GAMES,
            battle_progress * 100.0
        );
        // Print the progress text. The text is padded with spaces and ended with
        // a carriage return so old progress text is overwritten with new text.
        // Also, the standard output is flushed so the user sees the text we
        // printed instead of it getting stuck in the buffer.
        print!("{:50}\r", progress_text);
        let _ignored_result = io::stdout().flush();
        *last_update_time = time::Instant::now();
    }
}

// Gets the name of a provided AI difficulty.
fn get_difficulty_name(difficulty: &ai::Difficulty) -> &str {
    match difficulty {
        ai::Difficulty::None => "None",
        ai::Difficulty::Easy => "Easy",
        ai::Difficulty::Medium => "Medium",
        ai::Difficulty::Hard => "Hard",
        ai::Difficulty::Unbeatable => "Unbeatable",
        ai::Difficulty::Custom(_) => "Custom",
    }
}

// Holds the battle's scores and provides convenience methods for calculating
// the percentage of wins or cats games.
struct BattleScores {
    player_x_wins: i32,
    player_o_wins: i32,
    cats_games: i32,
}

impl BattleScores {
    fn new() -> Self {
        BattleScores {
            player_x_wins: 0,
            player_o_wins: 0,
            cats_games: 0,
        }
    }

    fn total_games(&self) -> i32 {
        self.player_x_wins + self.player_o_wins + self.cats_games
    }

    fn player_x_win_percent(&self) -> f64 {
        self.calculate_percent(self.player_x_wins)
    }

    fn player_o_win_percent(&self) -> f64 {
        self.calculate_percent(self.player_o_wins)
    }

    fn cats_game_percent(&self) -> f64 {
        self.calculate_percent(self.cats_games)
    }

    fn calculate_percent(&self, value: i32) -> f64 {
        if self.total_games() > 0 {
            let fraction = value as f64 / self.total_games() as f64;
            fraction * 100.0
        } else {
            0.0
        }
    }
}

impl fmt::Display for BattleScores {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{:3.0}% - {:3.0}% - {:3.0}%",
            self.player_x_win_percent(),
            self.player_o_win_percent(),
            self.cats_game_percent()
        )?;
        Ok(())
    }
}
